chainstate/tx-verifier/src/transaction_verifier/input_output_policy.rs

```diff
@@ -13,75 +13,778 @@
+                    SpendStakeError::ConsensusPoSError(ConsensusPoSError::NoKernel),
+                )),
+                // single input
+                [intput_utxo] => match intput_utxo {
+                    TxOutput::Transfer(_, _)
+                    | TxOutput::LockThenTransfer(_, _, _)
+                    | TxOutput::Burn(_)
+                    | TxOutput::DecommissionPool(_, _, _, _) => {
+                        Err(ConnectTransactionError::InvalidInputTypeInReward)
+                    }
+                    TxOutput::StakePool(_) | TxOutput::ProduceBlockFromStake(_, _, _) => {
+                        let outputs =
+                            reward.outputs().ok_or(ConnectTransactionError::SpendStakeError(
+                                SpendStakeError::NoBlockRewardOutputs,
+                            ))?;
+                        match outputs {
+                            [] => Err(ConnectTransactionError::SpendStakeError(
+                                SpendStakeError::NoBlockRewardOutputs,
+                            )),
+                            [output] => match output {
+                                TxOutput::Transfer(_, _)
+                                | TxOutput::LockThenTransfer(_, _, _)
+                                | TxOutput::Burn(_)
+                                | TxOutput::StakePool(_)
+                                | TxOutput::DecommissionPool(_, _, _, _) => {
+                                    Err(ConnectTransactionError::InvalidOutputTypeInReward)
+                                }
+                                TxOutput::ProduceBlockFromStake(_, _, _) => Ok(()),
+                            },
+                            _ => Err(ConnectTransactionError::SpendStakeError(
+                                SpendStakeError::MultipleBlockRewardOutputs,
+                            )),
+                        }
+                    }
+                },
+                // multiple inputs
+                _ => Err(ConnectTransactionError::SpendStakeError(
+                    SpendStakeError::ConsensusPoSError(ConsensusPoSError::MultipleKernels),
+                )),
+            }
+        }
+        None => {
+            // if no kernel input is present it's allowed to have multiple `LockThenTransfer` outputs
+            let all_lock_then_transfer = reward
+                .outputs()
+                .ok_or(ConnectTransactionError::SpendStakeError(
+                    SpendStakeError::NoBlockRewardOutputs,
+                ))?
+                .iter()
+                .all(|output| match output {
+                    TxOutput::LockThenTransfer(_, _, _) => true,
+                    TxOutput::Transfer(_, _)
+                    | TxOutput::Burn(_)
+                    | TxOutput::StakePool(_)
+                    | TxOutput::ProduceBlockFromStake(_, _, _)
+                    | TxOutput::DecommissionPool(_, _, _, _) => false,
+                });
```

Comment:
Are there supposed to be any restrictions on the time lock duration?

Reply:
Output timelocks are checked in `ChainstateRef::check_outputs_timelock`